CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc

```diff
@@ -0,0 +1,189 @@
+// authors: Jan Kaspar (jan.kaspar@gmail.com)
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ModuleFactory.h"
+#include "FWCore/Framework/interface/ESProducer.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "CondFormats/DataRecord/interface/LHCInfoRcd.h"
+#include "CondFormats/DataRecord/interface/CTPPSOpticsRcd.h"
+#include "CondFormats/DataRecord/interface/CTPPSInterpolatedOpticsRcd.h"
+
+#include "CondFormats/RunInfo/interface/LHCInfo.h"
+#include "CondFormats/CTPPSReadoutObjects/interface/LHCOpticalFunctionsSetCollection.h"
+#include "CondFormats/CTPPSReadoutObjects/interface/LHCInterpolatedOpticalFunctionsSetCollection.h"
+
+
+class CTPPSInterpolatedOpticalFunctionsESSource : public edm::ESProducer
+{
+  public:
+    CTPPSInterpolatedOpticalFunctionsESSource(const edm::ParameterSet&);
+    ~CTPPSInterpolatedOpticalFunctionsESSource() override {}
+
+    std::shared_ptr<LHCInterpolatedOpticalFunctionsSetCollection> produce(const CTPPSInterpolatedOpticsRcd&);
```

Comment:
The output data are buffered:
https://github.com/CTPPS/cmssw/blob/proton_reco_10_4_X/CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc#L31
so there is sharing of the data between the ESSource and the ES framework. Isn't the shared_ptr more appropriate in this case?

Reply:
@makortel please check.
My comment to use a unique_ptr was based on the recent migrations in various ES producers

Reply:
We are working to allow multiple concurrent IOVs. If you use a shared_ptr, the same object could be used in both IOVs where we are supposed to have different objects (i.e. you'll wind up effectively changing the 1st IOVs data product).

If that is not the case here, i.e. instead you are incorrectly telling the framework that an IOV should change but, in fact, it shouldn't have since the data objects are the same and you make sure to always make a new object instead of modifying the existing one then using a shared_ptr is OK.

Reply:
I'm not sure I understand. What is happening here is that we want to change the data (optical functions) only when the crossing angle (xangle) changes. The xangle is part of the LHCInfo. Some fields of the LHCInfo change frequently, but the xangle changes less often. Therefore the current data are buffered and are reused until a true change of xangle occurs.